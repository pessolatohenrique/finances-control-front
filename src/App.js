import React from "react";
import AppBarWrapper from "./components/AppBarWrapper";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import ExampleList from "./views/register/ExampleList";
import ExampleForm from "./views/register/ExampleForm";
import ExampleDetail from "./views/register/ExampleDetail";
import NotFound from "./components/NotFound";
import LoginForm from "./views/register/LoginForm";

function App() {
  return (
    <Router>
      <AppBarWrapper />

      <Switch>
        <Route exact path="/signin">
          <LoginForm />
        </Route>

        <Route exact path="/livros">
          <ExampleList />
        </Route>

        <Route exact path="/autores/novo">
          <ExampleForm />
        </Route>

        <Route exact path="/livros/:id">
          <ExampleDetail />
        </Route>

        <Route path="*">
          <NotFound />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
