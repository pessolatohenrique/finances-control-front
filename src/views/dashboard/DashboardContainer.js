import React, { useState, useEffect } from "react";
import axios from "axios";
import moment from "moment";
import {
  Grid,
  Card,
  CardContent,
  CardActions,
  CardMedia,
  Typography,
  Alert,
  Button,
  Snackbar,
  TextField,
} from "@mui/material";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { DatePicker } from "@mui/x-date-pickers/DatePicker";
import ViewListToggle from "../../components/ViewListToggle";
import { EarningTable } from "../earnings/EarningTable";
import { EarningList } from "../earnings/EarningList";
import { ExpenseTable } from "../expenses/ExpenseTable";
import { ExpenseList } from "../expenses/ExpenseList";
import { THEME_COLOR } from "../../constants/default_settings";
import useView from "../../hooks/useView";
import earningImage from "../../assets/earnings.jpg";
import expensesImage from "../../assets/expenses.jpg";
import indicatorsImage from "../../assets/indicators.png";

import {
  SNACKBAR_DIRECTION,
  DATE_MIN_FILTER,
  DATE_MAX_FILTER,
} from "../../constants/default_settings";
import useToast from "../../hooks/useToast";

function createData(name, author, pages, genre) {
  return { name, author, pages, genre };
}

const rows = [
  createData("Jogos Vorazes", "Suzanne Collins", 320, "Trilogia"),
  createData(
    "Harry Potter e a Pedra Filosofal",
    "J.K Rowling",
    220,
    "Trilogia"
  ),
  createData("Me poupe!", "Nathalia Arcuri", 200, "Finanças"),
  createData("Sherlock Homes", "Arthur Conan Doyle", 500, "Suspense"),
  createData(
    "O mistério dos sete relógios",
    "Agatha Christie",
    300,
    "Suspense"
  ),
];

function useViewEarning() {
  const [isTable, isList, switchFormat] = useView();
  return [isTable, isList, switchFormat];
}

function useViewExpense() {
  const [isTable, isList, switchFormat] = useView();
  return [isTable, isList, switchFormat];
}

function IndicatorCard({ image, title, subtitle }) {
  return (
    <Card sx={{ width: "27%" }} data-testid="indicator-card">
      <CardMedia
        component="img"
        height="140"
        image={image}
        alt="green iguana"
        sx={{ objectFit: "initial" }}
      />
      <CardContent>
        <Typography variant="h5" component="div">
          {title}
        </Typography>
        <Typography variant="body2" color="text.secondary">
          {subtitle}
        </Typography>
      </CardContent>
    </Card>
  );
}

function DashboardContainer() {
  // custom hooks
  const { open, error, setError, showToast, hideToast } = useToast();
  const [isTableEarning, isListEarning, switchFormatEarning] = useViewEarning();
  const [isTableExpense, isListExpense, switchFormatExpense] = useViewEarning();

  // local states
  const [userRecipe, setUserRecipe] = useState();
  const [budget, setBudget] = useState();
  const [monthFilter, setMonthFilter] = useState(moment().format("MM"));
  const [yearFilter, setYearFilter] = useState(moment().format("YYYY"));
  const [fullDate, setFullDate] = useState(new Date());

  console.log("budget", budget);

  useEffect(() => {
    async function getUserRecipe() {
      try {
        const response = await axios.get(
          `budget/summarize?month=${monthFilter}&year=${yearFilter}`
        );
        setBudget(response.data);
      } catch (error) {
        showToast();
        setError(error?.response?.data?.message || null);
      }
    }

    async function getBudget() {
      try {
        const response = await axios.get(`/user/recipe`);
        setUserRecipe(response.data);
      } catch (error) {
        showToast();
        setError(error?.response?.data?.message || null);
      }
    }

    getUserRecipe();
    getBudget();
  }, [monthFilter]);

  function updateDateFilter(dateParam) {
    setFullDate(dateParam);
    setMonthFilter(moment(dateParam).format("MM"));
    setYearFilter(moment(dateParam).format("YYYY"));
  }

  return (
    <Grid
      minHeight="100vh"
      sx={{
        backgroundColor: `#E5E5E5`,
        overflow: "hidden",
      }}
    >
      <br />
      <Snackbar
        anchorOrigin={SNACKBAR_DIRECTION}
        open={open}
        autoHideDuration={6000}
        onClose={hideToast}
      >
        <Alert onClose={hideToast} severity="error" sx={{ width: "100%" }}>
          {error}
        </Alert>
      </Snackbar>

      <Grid sx={{ marginLeft: 15, marginRight: 15, marginBottom: 2 }}>
        <LocalizationProvider dateAdapter={AdapterDateFns}>
          <DatePicker
            label="Selecione a data"
            inputProps={{ "data-testid": "dateFilter" }}
            views={["year", "month"]}
            minDate={DATE_MIN_FILTER}
            maxDate={DATE_MAX_FILTER}
            inputFormat="MM/yyyy"
            value={fullDate}
            onChange={(dateParam) => {
              updateDateFilter(dateParam);
            }}
            renderInput={(params) => (
              <TextField
                inputProps={{ "data-testid": "dateFilter" }}
                {...params}
              />
            )}
          />
        </LocalizationProvider>
      </Grid>

      <Grid
        display="flex"
        flexDirection="row"
        justifyContent="space-between"
        sx={{ marginLeft: 15, marginRight: 15 }}
      >
        {!userRecipe && (
          <Alert
            severity="warning"
            action={
              <Button
                color="inherit"
                size="small"
                href="/receita/associar"
                data-testid="associate-button"
              >
                <strong>{"Associar".toUpperCase()}</strong>
              </Button>
            }
          >
            Para receber dicas personalizadas, associe uma "receita do sucesso"!
          </Alert>
        )}

        <IndicatorCard
          image={earningImage}
          title={`R$ ${budget?.sum_earning?.toFixed(2)}`}
          subtitle="de ganhos"
        />
        <IndicatorCard
          image={expensesImage}
          title={`R$ ${budget?.sum_expense?.toFixed(2)}`}
          subtitle="de gastos"
        />
        <IndicatorCard
          image={indicatorsImage}
          title={`${budget?.sum_percentage?.toFixed(2)}%`}
          subtitle="de gastos da receita de sucesso"
        />
      </Grid>

      <Grid
        sx={{ marginLeft: 15, marginRight: 15, marginTop: 5, marginBottom: 5 }}
      >
        <Card>
          <CardContent>
            <Typography
              gutterBottom
              variant="h5"
              component="h1"
              color={THEME_COLOR}
            >
              Ganhos
            </Typography>

            <ViewListToggle
              isTable={isTableEarning}
              isList={isListEarning}
              switchFormat={switchFormatEarning}
            />

            {isTableEarning() && <EarningTable budget={budget} />}

            {isListEarning() && <EarningList budget={budget} />}
          </CardContent>
          <CardActions>
            <Button size="small" href="/autores/novo">
              Adicionar
            </Button>
          </CardActions>
        </Card>
      </Grid>

      <Grid
        sx={{ marginLeft: 15, marginRight: 15, marginTop: 5, marginBottom: 5 }}
      >
        <Card>
          <CardContent>
            <Typography
              gutterBottom
              variant="h5"
              component="h1"
              color={THEME_COLOR}
            >
              Despesas
            </Typography>

            <ViewListToggle
              isTable={isTableExpense}
              isList={isListExpense}
              switchFormat={switchFormatExpense}
            />

            {isTableExpense() && <ExpenseTable budget={budget} />}

            {isListExpense() && <ExpenseList budget={budget} />}
          </CardContent>
          <CardActions>
            <Button size="small" href="/autores/novo">
              Adicionar
            </Button>
          </CardActions>
        </Card>
      </Grid>
    </Grid>
  );
}

export default DashboardContainer;
