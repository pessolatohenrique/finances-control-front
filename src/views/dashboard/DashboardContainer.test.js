import {
  render,
  screen,
  fireEvent,
  waitForElement,
  waitForDomChange,
  waitFor,
} from "@testing-library/react";
import App from "../../App";
import DashboardContainer from "./DashboardContainer";

describe("Dashboard Container", () => {
  beforeAll(() => {
    console.log("JEST", process.env.JEST_WORKER_ID);

    render(<App />);
    const submitBtn = screen.getByTestId("submit-button");
    const username = screen.getByTestId("username");
    const password = screen.getByTestId("password");

    fireEvent.input(username, {
      target: { value: process.env.REACT_APP_API_DEFAULT_USER_WITHOUT_RECIPE },
    });
    fireEvent.input(password, {
      target: {
        value: process.env.REACT_APP_API_DEFAULT_PASSWORD_WITHOUT_RECIPE,
      },
    });
    fireEvent.click(submitBtn);
  });

  it("should generate snapshot", () => {
    const { container } = render(<DashboardContainer />);
    expect(container).toMatchSnapshot();
  });

  it("should show alert when don't have a recipe", async () => {
    render(<DashboardContainer />);
    const associateBtn = screen.getByTestId("associate-button");
    expect(associateBtn).toBeInTheDocument();
  });

  it("should not show alert when have a recipe", async () => {
    render(<App />);
    const submitBtn = screen.getByTestId("submit-button");
    const username = screen.getByTestId("username");
    const password = screen.getByTestId("password");

    fireEvent.input(username, {
      target: { value: process.env.REACT_APP_API_DEFAULT_USER },
    });
    fireEvent.input(password, {
      target: {
        value: process.env.REACT_APP_API_DEFAULT_PASSWORD,
      },
    });
    fireEvent.click(submitBtn);

    await waitFor(() =>
      expect(screen.queryByText("Associar")).not.toBeInTheDocument()
    );
  });

  it("should load indicators cards", async () => {
    render(<DashboardContainer />);
    const indicatorCard = screen.getAllByTestId("indicator-card");
    expect(indicatorCard.length).toBe(3);
  });
});
