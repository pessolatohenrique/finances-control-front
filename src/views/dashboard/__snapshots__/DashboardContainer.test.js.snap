// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Container should generate snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root css-4m4w48-MuiGrid-root"
  >
    <br />
    <div
      class="MuiGrid-root css-12lyjsw-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-1"
          id="mui-1-label"
        >
          Selecione a data
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            aria-label="Choose date, selected date is May 14, 2022"
            aria-readonly="true"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-1"
            readonly=""
            type="text"
            value="05/2022"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1in441m"
            >
              <span>
                Selecione a data
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root css-1mvjg4q-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-rob5aa-MuiPaper-root-MuiAlert-root"
        role="alert"
      >
        <div
          class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="ReportProblemOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-acap47-MuiAlert-message"
        >
          Para receber dicas personalizadas, associe uma "receita do sucesso"!
        </div>
        <div
          class="MuiAlert-action css-ki1hdl-MuiAlert-action"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorInherit MuiButtonBase-root  css-u44ztt-MuiButtonBase-root-MuiButton-root"
            data-testid="associate-button"
            href="/receita/associar"
            tabindex="0"
          >
            <strong>
              ASSOCIAR
            </strong>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1cllq7-MuiPaper-root-MuiCard-root"
        data-testid="indicator-card"
      >
        <img
          alt="green iguana"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-dzbxvc-MuiCardMedia-root"
          height="140"
          src="earnings.jpg"
        />
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            R$ undefined
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            de ganhos
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1cllq7-MuiPaper-root-MuiCard-root"
        data-testid="indicator-card"
      >
        <img
          alt="green iguana"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-dzbxvc-MuiCardMedia-root"
          height="140"
          src="expenses.jpg"
        />
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            R$ undefined
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            de gastos
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1cllq7-MuiPaper-root-MuiCard-root"
        data-testid="indicator-card"
      >
        <img
          alt="green iguana"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-dzbxvc-MuiCardMedia-root"
          height="140"
          src="indicators.png"
        />
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            undefined%
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            de gastos da receita de sucesso
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root css-h345q-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h1
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1y15g74-MuiTypography-root"
          >
            Ganhos
          </h1>
          <div
            class="MuiGrid-root MuiGrid-container css-1mx4t5n-MuiGrid-root"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="view-table"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ListIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="view-card"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ViewModuleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M4 11h5V5H4v6zm0 7h5v-6H4v6zm6 0h5v-6h-5v6zm6 0h5v-6h-5v6zm-6-7h5V5h-5v6zm6-6v6h5V5h-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
          >
            <table
              class="MuiTable-root css-rqglhn-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1w4f6gj-MuiTableCell-root"
                    scope="col"
                  >
                    Descrição
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1w4f6gj-MuiTableCell-root"
                    scope="col"
                  >
                    Valor
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1w4f6gj-MuiTableCell-root"
                    scope="col"
                  >
                    Data
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              />
            </table>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            href="/autores/novo"
            tabindex="0"
          >
            Adicionar
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
